src/Ema/Route/CtxPrism.hs

```diff
@@ -0,0 +1,70 @@
+module Ema.Route.CtxPrism
+  ( -- * Type
+    CtxPrism,
+
+    -- * Construction
+    fromPrism,
+
+    -- * Conversion
+    cpreview,
+    creview,
+
+    -- * Functor
+    cpmap,
+
+    -- * Law checks
+    ctxPrismIsLawfulFor,
+  )
+where
+
+import Control.Monad.Writer (Writer, tell)
+import Optics.Core
+  ( Prism',
+    preview,
+    prism',
+    review,
+    (%),
+  )
+
+-- | A `Prism` with a context
+-- This can't actually be a prism due to coercion problems. Use `toPrism` & `fromPrism`.
+-- See https://stackoverflow.com/q/71489589/55246
+type CtxPrism ctx s a =
+  -- FIXME: ought to be `ctx -> Prism' s a`
+  ctx -> (a -> s, s -> Maybe a)
```

Comment:
I suspect the `optics-core` way would be to do something more like

```haskell
type CtxOptic' bleep bloop ctx s a = ctx -> Optic' bleep bloop s a
```

Then you can put the specific constraints you actually need on `bleep` and `bloop` in each situation, rather than assuming an unindexed prism.

Reply:
@treeowl Unfortunately I cannot use optics types (monomorphic or polymorphic) here because not all constructors are exported to be possible for me to do coercion-based DerivingVia. See https://stackoverflow.com/q/71489589/55246

This is why the representation remains a function here.